Error: top-level heading or title meta information must be set

Hello there 👋
I got following error:
```java
Exception in thread "main" java.lang.RuntimeException: failed to create confluence page for asciidoc content in '***'
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluencePage.newAsciidocConfluencePage(AsciidocConfluencePage.java:159)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluenceConverter.lambda$buildPageTree$0(AsciidocConfluenceConverter.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1085)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluenceConverter.buildPageTree(AsciidocConfluenceConverter.java:91)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluenceConverter.lambda$buildPageTree$0(AsciidocConfluenceConverter.java:102)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluenceConverter.buildPageTree(AsciidocConfluenceConverter.java:91)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluenceConverter.convert(AsciidocConfluenceConverter.java:78)
	at org.sahli.asciidoc.confluence.publisher.cli.AsciidocConfluencePublisherCommandLineClient.main(AsciidocConfluencePublisherCommandLineClient.java:100)
Caused by: java.lang.RuntimeException: top-level heading or title meta information must be set
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluencePage.lambda$pageTitle$8(AsciidocConfluencePage.java:216)
	at java.base/java.util.Optional.orElseThrow(Optional.java:408)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluencePage.pageTitle(AsciidocConfluencePage.java:216)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluencePage.lambda$replaceCrossReferenceTargets$9(AsciidocConfluencePage.java:253)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluencePage.replaceAll(AsciidocConfluencePage.java:198)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluencePage.lambda$replaceCrossReferenceTargets$10(AsciidocConfluencePage.java:246)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluencePage.lambda$postProcessContent$4(AsciidocConfluencePage.java:208)
	at java.base/java.util.stream.ReduceOps$1ReducingSink.accept(ReduceOps.java:80)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.reduce(ReferencePipeline.java:563)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluencePage.postProcessContent(AsciidocConfluencePage.java:208)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluencePage.convertedContent(AsciidocConfluencePage.java:169)
	at org.sahli.asciidoc.confluence.publisher.converter.AsciidocConfluencePage.newAsciidocConfluencePage(AsciidocConfluencePage.java:153)
	... 9 more
```

This error occurs when I am not adding a top-level title directly, but including another template file which injects the toplevel title. This is probably caused by [this statement](https://github.com/confluence-publisher/confluence-publisher/blob/8219f4951942296ba9f0b80303a9df30f80efc66/asciidoc-confluence-publisher-converter/src/main/java/org/sahli/asciidoc/confluence/publisher/converter/AsciidocConfluencePage.java#L252) not acknowledging includes.

Kind regards,
Marvin Haase

--- a/asciidoc-confluence-publisher-converter/src/main/java/org/sahli/asciidoc/confluence/publisher/converter/AsciidocConfluencePage.java
+++ b/asciidoc-confluence-publisher-converter/src/main/java/org/sahli/asciidoc/confluence/publisher/converter/AsciidocConfluencePage.java
@@ -150,7 +150,7 @@ public class AsciidocConfluencePage {
             Document document = ASCIIDOCTOR.load(asciidocContent, options);
 
             String pageTitle = unescapeHtml(pageTitle(document, userAttributesWithMaskedNullValues, pageTitlePostProcessor));
-            String pageContent = convertedContent(document, asciidocPagePath, attachmentCollector, userAttributesWithMaskedNullValues, pageTitlePostProcessor, sourceEncoding, spaceKey);
+            String pageContent = convertedContent(document, asciidocPagePath, attachmentCollector, userAttributesWithMaskedNullValues, pageTitlePostProcessor, sourceEncoding, spaceKey, options);
 
             List<String> keywords = keywords(document);
 
@@ -164,10 +164,10 @@ public class AsciidocConfluencePage {
         return path.contains("/") ? path.substring(path.lastIndexOf('/') + 1) : path;
     }
 
-    private static String convertedContent(Document document, Path pagePath, Map<String, String> attachmentCollector, Map<String, Object> userAttributes, PageTitlePostProcessor pageTitlePostProcessor, Charset sourceEncoding, String spaceKey) {
+    private static String convertedContent(Document document, Path pagePath, Map<String, String> attachmentCollector, Map<String, Object> userAttributes, PageTitlePostProcessor pageTitlePostProcessor, Charset sourceEncoding, String spaceKey, Options options) {
         String content = document.convert();
         String postProcessedContent = postProcessContent(content,
-                replaceCrossReferenceTargets(pagePath, userAttributes, pageTitlePostProcessor, sourceEncoding, spaceKey),
+                replaceCrossReferenceTargets(pagePath, userAttributes, pageTitlePostProcessor, sourceEncoding, spaceKey, options),
                 collectAndReplaceAttachmentFileNames(attachmentCollector, sourceEncoding),
                 unescapeCdataHtmlContent()
         );
@@ -242,14 +242,15 @@ public class AsciidocConfluencePage {
                 .build();
     }
 
-    private static Function<String, String> replaceCrossReferenceTargets(Path pagePath, Map<String, Object> userAttributes, PageTitlePostProcessor pageTitlePostProcessor, Charset sourceEncoding, String spaceKey) {
+    private static Function<String, String> replaceCrossReferenceTargets(Path pagePath, Map<String, Object> userAttributes, PageTitlePostProcessor pageTitlePostProcessor, Charset sourceEncoding, String spaceKey, Options options) {
         return (content) -> replaceAll(content, PAGE_TITLE_PATTERN, (matchResult) -> {
             String htmlTarget = matchResult.group(1);
             Path referencedPagePath = pagePath.getParent().resolve(Paths.get(htmlTarget.substring(0, htmlTarget.lastIndexOf('.')) + ".adoc"));
 
             try {
+                options.setBaseDir(referencedPagePath.getParent().toString());
                 String referencedPageContent = readIntoString(new FileInputStream(referencedPagePath.toFile()), sourceEncoding);
-                Document referencedDocument = ASCIIDOCTOR.load(referencedPageContent, Options.builder().parseHeaderOnly(true).build());
+                Document referencedDocument = ASCIIDOCTOR.load(referencedPageContent, options);
                 String referencedPageTitle = pageTitle(referencedDocument, userAttributes, pageTitlePostProcessor);
 
                 /*

--- a/asciidoc-confluence-publisher-converter/src/test/java/org/sahli/asciidoc/confluence/publisher/converter/AsciidocConfluencePageTest.java
+++ b/asciidoc-confluence-publisher-converter/src/test/java/org/sahli/asciidoc/confluence/publisher/converter/AsciidocConfluencePageTest.java
@@ -1074,6 +1074,22 @@ public class AsciidocConfluencePageTest {
     }
 
     @Test
+    public void renderConfluencePage_asciiDocWithInterDocumentCrossReference_returnsConfluencePageWithLinkToReferencedPageByPageTitleAcknowledgingIncludes() {
+        // arrange
+        Path rootFolder = copyAsciidocSourceToTemporaryFolder("src/test/resources/inter-document-cross-references");
+        AsciidocPage asciidocPage = asciidocPage(rootFolder, "source-page-with-title-in-include.adoc");
+
+        // act
+        AsciidocConfluencePage asciidocConfluencePage = newAsciidocConfluencePage(asciidocPage, UTF_8, TEMPLATES_FOLDER, assetsTargetFolderFor(asciidocPage), "TEST");
+
+        // assert
+        String expectedContent = "<p>This is a <ac:link><ri:page ri:content-title=\"Target Page\" ri:space-key=\"TEST\"></ri:page>" +
+                "<ac:plain-text-link-body><![CDATA[reference]]></ac:plain-text-link-body>" +
+                "</ac:link> to the target page.</p>";
+        assertThat(asciidocConfluencePage.content(), is(expectedContent));
+    }
+
+    @Test
     public void renderConfluencePage_asciiDocWithInterDocumentCrossReferenceAnchor_returnsConfluencePageWithLinkToReferencedPageByPageTitleWithAnchor() {
         // arrange
         Path rootFolder = copyAsciidocSourceToTemporaryFolder("src/test/resources/inter-document-cross-references");

--- a/asciidoc-confluence-publisher-converter/src/test/java/org/sahli/asciidoc/confluence/publisher/converter/AsciidocConfluencePageTest.java
+++ b/asciidoc-confluence-publisher-converter/src/test/java/org/sahli/asciidoc/confluence/publisher/converter/AsciidocConfluencePageTest.java
@@ -1073,7 +1073,6 @@ public class AsciidocConfluencePageTest {
         assertThat(asciidocConfluencePage.content(), is(expectedContent));
     }
 
-    // This is basically just a check if the base-dir was set correctly in the options instance used to parse the referenced page
     @Test
     public void renderConfluencePage_asciiDocWithInterDocumentCrossReference_returnsConfluencePageWithLinkToReferencedPageByPageTitleAcknowledgingIncludes() {
         // arrange
@@ -1090,6 +1089,7 @@ public class AsciidocConfluencePageTest {
         assertThat(asciidocConfluencePage.content(), is(expectedContent));
     }
 
+    // This is basically just a check if the base-dir was set correctly in the options instance used to parse the referenced page
     @Test
     public void renderConfluencePage_asciiDocWithInterDocumentCrossReferenceInSubDirectory_returnsConfluencePageWithLinkToReferencedPageByPageTitleAcknowledgingIncludes() {
         // arrange

--- a/asciidoc-confluence-publisher-converter/src/test/java/org/sahli/asciidoc/confluence/publisher/converter/AsciidocConfluencePageTest.java
+++ b/asciidoc-confluence-publisher-converter/src/test/java/org/sahli/asciidoc/confluence/publisher/converter/AsciidocConfluencePageTest.java
@@ -1073,11 +1073,28 @@ public class AsciidocConfluencePageTest {
         assertThat(asciidocConfluencePage.content(), is(expectedContent));
     }
 
+    // This is basically just a check if the base-dir was set correctly in the options instance used to parse the referenced page
     @Test
     public void renderConfluencePage_asciiDocWithInterDocumentCrossReference_returnsConfluencePageWithLinkToReferencedPageByPageTitleAcknowledgingIncludes() {
         // arrange
         Path rootFolder = copyAsciidocSourceToTemporaryFolder("src/test/resources/inter-document-cross-references");
         AsciidocPage asciidocPage = asciidocPage(rootFolder, "source-page-with-title-in-include.adoc");
+
+        // act
+        AsciidocConfluencePage asciidocConfluencePage = newAsciidocConfluencePage(asciidocPage, UTF_8, TEMPLATES_FOLDER, assetsTargetFolderFor(asciidocPage), "TEST");
+
+        // assert
+        String expectedContent = "<p>This is a <ac:link><ri:page ri:content-title=\"Target Page\" ri:space-key=\"TEST\"></ri:page>" +
+                "<ac:plain-text-link-body><![CDATA[reference]]></ac:plain-text-link-body>" +
+                "</ac:link> to the target page.</p>";
+        assertThat(asciidocConfluencePage.content(), is(expectedContent));
+    }
+
+    @Test
+    public void renderConfluencePage_asciiDocWithInterDocumentCrossReferenceInSubDirectory_returnsConfluencePageWithLinkToReferencedPageByPageTitleAcknowledgingIncludes() {
+        // arrange
+        Path rootFolder = copyAsciidocSourceToTemporaryFolder("src/test/resources/inter-document-cross-references");
+        AsciidocPage asciidocPage = asciidocPage(rootFolder, "source-page-with-title-in-include-in-subdir.adoc");
 
         // act
         AsciidocConfluencePage asciidocConfluencePage = newAsciidocConfluencePage(asciidocPage, UTF_8, TEMPLATES_FOLDER, assetsTargetFolderFor(asciidocPage), "TEST");

--- a/asciidoc-confluence-publisher-converter/src/test/resources/inter-document-cross-references/include-page-with-title-in-include.adoc
+++ b/asciidoc-confluence-publisher-converter/src/test/resources/inter-document-cross-references/include-page-with-title-in-include.adoc
@@ -1,0 +1,1 @@
+= {title-key}

--- a/asciidoc-confluence-publisher-converter/src/test/resources/inter-document-cross-references/include-page-with-title-in-include.adoc
+++ b/asciidoc-confluence-publisher-converter/src/test/resources/inter-document-cross-references/include-page-with-title-in-include.adoc
@@ -1,1 +1,1 @@
-= {title-key}
+= Target Page

--- a/asciidoc-confluence-publisher-converter/src/test/resources/inter-document-cross-references/source-page-with-title-in-include-in-subdir.adoc
+++ b/asciidoc-confluence-publisher-converter/src/test/resources/inter-document-cross-references/source-page-with-title-in-include-in-subdir.adoc
@@ -1,0 +1,3 @@
+= Source Page
+
+This is a xref:subdir/target-page-with-title-in-include-relative-to-this-file.adoc[reference] to the target page.

--- a/asciidoc-confluence-publisher-converter/src/test/resources/inter-document-cross-references/source-page-with-title-in-include.adoc
+++ b/asciidoc-confluence-publisher-converter/src/test/resources/inter-document-cross-references/source-page-with-title-in-include.adoc
@@ -1,0 +1,3 @@
+= Source Page
+
+This is a xref:target-page-with-title-in-include.adoc[reference] to the target page.

--- a/asciidoc-confluence-publisher-converter/src/test/resources/inter-document-cross-references/subdir/target-page-with-title-in-include-relative-to-this-file.adoc
+++ b/asciidoc-confluence-publisher-converter/src/test/resources/inter-document-cross-references/subdir/target-page-with-title-in-include-relative-to-this-file.adoc
@@ -1,0 +1,1 @@
+include::../include-page-with-title-in-include.adoc[]

--- a/asciidoc-confluence-publisher-converter/src/test/resources/inter-document-cross-references/target-page-with-title-in-include.adoc
+++ b/asciidoc-confluence-publisher-converter/src/test/resources/inter-document-cross-references/target-page-with-title-in-include.adoc
@@ -1,2 +1,1 @@
-:title-key: Target Page
 include::include-page-with-title-in-include.adoc[]

--- a/asciidoc-confluence-publisher-converter/src/test/resources/inter-document-cross-references/target-page-with-title-in-include.adoc
+++ b/asciidoc-confluence-publisher-converter/src/test/resources/inter-document-cross-references/target-page-with-title-in-include.adoc
@@ -1,0 +1,2 @@
+:title-key: Target Page
+include::include-page-with-title-in-include.adoc[]

